binding/form_mapping.go

```diff
@@ -20,121 +24,156 @@ func mapForm(ptr interface{}, form map[string][]string) error {
-			if defaultList[0] == "default" {
-				defaultValue = defaultList[1]
-			}
+func mapping(value reflect.Value, field reflect.StructField, form map[string][]string, tag string) (bool, error) {
+	var vKind = value.Kind()
+
+	// Ptr
+	if vKind == reflect.Ptr {
+		var isNew bool
+		vPtr := value
+		if value.IsNil() {
+			isNew = true
+			vPtr = reflect.New(value.Type().Elem())
 		}
-		if inputFieldName == "" {
-			inputFieldName = typeField.Name
-
-			// if "form" tag is nil, we inspect if the field is a struct or struct pointer.
-			// this would not make sense for JSON parsing but it does for a form
-			// since data is flatten
-			if structFieldKind == reflect.Ptr {
-				if !structField.Elem().IsValid() {
-					structField.Set(reflect.New(structField.Type().Elem()))
-				}
-				structField = structField.Elem()
-				structFieldKind = structField.Kind()
-			}
-			if structFieldKind == reflect.Struct {
-				err := mapFormByTag(structField.Addr().Interface(), form, tag)
-				if err != nil {
-					return err
-				}
-				continue
-			}
+		isSetted, err := mapping(vPtr.Elem(), field, form, tag)
+		if err != nil {
+			return false, err
+		}
+		if isNew && isSetted {
+			value.Set(vPtr)
 		}
-		inputValue, exists := form[inputFieldName]
+		return isSetted, nil
+	}
 
-		if !exists {
-			if defaultValue == "" {
+	// Try to set value
+	ok, err := tryToSetValue(value, field, form, tag)
+	if err != nil {
+		return false, err
+	}
+	if ok {
+		return true, nil
+	}
+
+	// Struct
+	if vKind == reflect.Struct {
+		tValue := value.Type()
+
+		var isSetted bool
+		for i := 0; i < value.NumField(); i++ {
+			if !value.Field(i).CanSet() {
 				continue
 			}
-			inputValue = make([]string, 1)
-			inputValue[0] = defaultValue
+			ok, err := mapping(value.Field(i), tValue.Field(i), form, tag)
+			if err != nil {
+				return false, err
+			}
+			isSetted = isSetted || ok
 		}
+		return isSetted, nil
+	}
+	return false, nil
+}
 
-		numElems := len(inputValue)
-		if structFieldKind == reflect.Slice && numElems > 0 {
-			sliceOf := structField.Type().Elem().Kind()
-			slice := reflect.MakeSlice(structField.Type(), numElems, numElems)
-			for i := 0; i < numElems; i++ {
-				if err := setWithProperType(sliceOf, inputValue[i], slice.Index(i)); err != nil {
-					return err
-				}
-			}
-			val.Field(i).Set(slice)
-			continue
+func tryToSetValue(value reflect.Value, field reflect.StructField, form map[string][]string, tag string) (bool, error) {
+	var tagValue, defaultValue string
+	var isDefaultExists bool
+
+	tagValue = field.Tag.Get(tag)
+	tagValue, opts := head(tagValue, ",")
+
+	if tagValue == "-" { // just ignoring this field
+		return false, nil
+	}
+	if tagValue == "" { // default value is FieldName
+		tagValue = field.Name
+	}
+	if tagValue == "" { // when field is "emptyField" variable
+		return false, nil
+	}
+
+	var opt string
+	for len(opts) > 0 {
+		opt, opts = head(opts, ",")
+
+		k, v := head(opt, "=")
+		switch k {
+		case "default":
```

Comment:
why use switch not if?

Reply:
No difference. But if in future you are decide to add extra param, just write
```
case "something":
  newParam = v
```